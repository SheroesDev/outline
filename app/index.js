// @flow
import * as React from 'react';
import { render } from 'react-dom';
import { Provider } from 'mobx-react';
import { ThemeProvider } from 'styled-components';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect,
} from 'react-router-dom';

import stores from 'stores';
import theme from 'shared/styles/theme';
import globalStyles from 'shared/styles/globals';
import 'shared/styles/prism.css';

import Home from 'scenes/Home';
import Dashboard from 'scenes/Dashboard';
import Starred from 'scenes/Starred';
import Drafts from 'scenes/Drafts';
import Collection from 'scenes/Collection';
import Document from 'scenes/Document';
import Search from 'scenes/Search';
import Settings from 'scenes/Settings';
import Details from 'scenes/Settings/Details';
import People from 'scenes/Settings/People';
import Slack from 'scenes/Settings/Slack';
import Shares from 'scenes/Settings/Shares';
import Tokens from 'scenes/Settings/Tokens';
import Export from 'scenes/Settings/Export';
import Error404 from 'scenes/Error404';

import ErrorBoundary from 'components/ErrorBoundary';
import ScrollToTop from 'components/ScrollToTop';
import ScrollToAnchor from 'components/ScrollToAnchor';
import Layout from 'components/Layout';
import Auth from 'components/Auth';
import RouteSidebarHidden from 'components/RouteSidebarHidden';

import { matchDocumentSlug } from 'utils/routeHelpers';

let DevTools;
if (__DEV__) {
  DevTools = require('mobx-react-devtools').default; // eslint-disable-line global-require
}

const notFoundSearch = () => <Search notFound />;
const DocumentNew = () => <Document newDocument />;
const RedirectDocument = ({ match }: { match: Object }) => (
  <Redirect to={`/doc/${match.params.documentSlug}`} />
);

globalStyles();

const element = document.getElementById('root');

if (element) {
  render(
    <React.Fragment>
      <ThemeProvider theme={theme}>
        <ErrorBoundary>
          <Provider {...stores}>
            <Router>
              <ScrollToTop>
                <ScrollToAnchor>
                  <Switch>
                    <Route exact path="/" component={Home} />
                    <Route exact path="/share/:shareId" component={Document} />
                    <Auth>
                      <Layout>
                        <Switch>
                          <Route
                            exact
                            path="/dashboard"
                            component={Dashboard}
                          />
                          <Route exact path="/starred" component={Starred} />
                          <Route exact path="/drafts" component={Drafts} />
                          <Route exact path="/settings" component={Settings} />
                          <Route
                            exact
                            path="/settings/details"
                            component={Details}
                          />
                          <Route
                            exact
                            path="/settings/people"
                            component={People}
                          />
                          <Route
                            exact
                            path="/settings/shares"
                            component={Shares}
                          />
                          <Route
                            exact
                            path="/settings/tokens"
                            component={Tokens}
                          />
                          <Route
                            exact
                            path="/settings/integrations/slack"
                            component={Slack}
                          />
                          <Route
                            exact
                            path="/settings/export"
                            component={Export}
                          />
                          <Route
                            exact
                            path="/collections/:id"
                            component={Collection}
                          />
                          <Route
                            exact
                            path={`/d/${matchDocumentSlug}`}
                            component={RedirectDocument}
                          />
                          <Route
                            exact
                            path={`/doc/${matchDocumentSlug}`}
                            component={Document}
                          />
                          <Route
                            exact
                            path={`/doc/${matchDocumentSlug}/move`}
                            component={Document}
                          />
                          <Route exact path="/search" component={Search} />
                          <Route
                            exact
                            path="/search/:query"
                            component={Search}
                          />
                          <Route path="/404" component={Error404} />
                          <RouteSidebarHidden
                            exact
                            path={`/doc/${matchDocumentSlug}/edit`}
                            component={Document}
                          />
                          <RouteSidebarHidden
                            exact
                            path="/collections/:id/new"
                            component={DocumentNew}
                          />
                          <Route component={notFoundSearch} />
                        </Switch>
                      </Layout>
                    </Auth>
                  </Switch>
                </ScrollToAnchor>
              </ScrollToTop>
            </Router>
          </Provider>
        </ErrorBoundary>
      </ThemeProvider>
      {DevTools && <DevTools position={{ bottom: 0, right: 0 }} />}
    </React.Fragment>,
    element
  );
}

window.addEventListener('load', async () => {
  // installation does not use Google Analytics, or tracking is blocked on client
  // no point loading the rest of the analytics bundles
  if (!process.env.GOOGLE_ANALYTICS_ID || !window.ga) return;

  // https://github.com/googleanalytics/autotrack/issues/137#issuecomment-305890099
  await import('autotrack/autotrack.js');

  window.ga('require', 'outboundLinkTracker');
  window.ga('require', 'urlChangeTracker');
});
