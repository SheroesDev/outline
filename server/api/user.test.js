/* eslint-disable flowtype/require-valid-file-annotation */
import TestServer from 'fetch-test-server';
import app from '..';

import { flushdb, seed } from '../test/support';
import { buildUser } from '../test/factories';

const server = new TestServer(app.callback());

beforeEach(flushdb);
afterAll(server.close);

describe('#user.info', async () => {
  it('should return known user', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.info', {
      body: { token: user.getJwtToken() },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(user.id);
    expect(body.data.name).toEqual(user.name);
  });

  it('should require authentication', async () => {
    const res = await server.post('/api/user.info');
    expect(res.status).toEqual(401);
  });
});

describe('#user.delete', async () => {
  it('should not allow deleting without confirmation', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.delete', {
      body: { token: user.getJwtToken() },
    });
    expect(res.status).toEqual(400);
  });

  it('should allow deleting last admin if only user', async () => {
    const user = await buildUser({ isAdmin: true });
    const res = await server.post('/api/user.delete', {
      body: { token: user.getJwtToken(), confirmation: true },
    });
    expect(res.status).toEqual(200);
  });

  it('should not allow deleting last admin if many users', async () => {
    const user = await buildUser({ isAdmin: true });
    await buildUser({ teamId: user.teamId, isAdmin: false });

    const res = await server.post('/api/user.delete', {
      body: { token: user.getJwtToken(), confirmation: true },
    });
    expect(res.status).toEqual(400);
  });

  it('should allow deleting user account with confirmation', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.delete', {
      body: { token: user.getJwtToken(), confirmation: true },
    });
    expect(res.status).toEqual(200);
  });

  it('should require authentication', async () => {
    const res = await server.post('/api/user.delete');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });
});

describe('#user.update', async () => {
  it('should update user profile information', async () => {
    const { user } = await seed();
    const res = await server.post('/api/user.update', {
      body: { token: user.getJwtToken(), name: 'New name' },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it('should require authentication', async () => {
    const res = await server.post('/api/user.update');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });
});

describe('#user.promote', async () => {
  it('should promote a new admin', async () => {
    const { admin, user } = await seed();

    const res = await server.post('/api/user.promote', {
      body: { token: admin.getJwtToken(), id: user.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it('should require admin', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.promote', {
      body: { token: user.getJwtToken(), id: user.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(403);
    expect(body).toMatchSnapshot();
  });
});

describe('#user.demote', async () => {
  it('should demote an admin', async () => {
    const { admin, user } = await seed();
    await user.update({ isAdmin: true }); // Make another admin

    const res = await server.post('/api/user.demote', {
      body: {
        token: admin.getJwtToken(),
        id: user.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it("shouldn't demote admins if only one available ", async () => {
    const admin = await buildUser({ isAdmin: true });

    const res = await server.post('/api/user.demote', {
      body: {
        token: admin.getJwtToken(),
        id: admin.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(400);
    expect(body).toMatchSnapshot();
  });

  it('should require admin', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.promote', {
      body: { token: user.getJwtToken(), id: user.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(403);
    expect(body).toMatchSnapshot();
  });
});

describe('#user.suspend', async () => {
  it('should suspend an user', async () => {
    const { admin, user } = await seed();

    const res = await server.post('/api/user.suspend', {
      body: {
        token: admin.getJwtToken(),
        id: user.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it("shouldn't allow suspending the user themselves", async () => {
    const admin = await buildUser({ isAdmin: true });
    const res = await server.post('/api/user.suspend', {
      body: {
        token: admin.getJwtToken(),
        id: admin.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(400);
    expect(body).toMatchSnapshot();
  });

  it('should require admin', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.suspend', {
      body: { token: user.getJwtToken(), id: user.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(403);
    expect(body).toMatchSnapshot();
  });
});

describe('#user.activate', async () => {
  it('should activate a suspended user', async () => {
    const { admin, user } = await seed();
    await user.update({
      suspendedById: admin.id,
      suspendedAt: new Date(),
    });

    expect(user.isSuspended).toBe(true);
    const res = await server.post('/api/user.activate', {
      body: {
        token: admin.getJwtToken(),
        id: user.id,
      },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body).toMatchSnapshot();
  });

  it('should require admin', async () => {
    const user = await buildUser();
    const res = await server.post('/api/user.activate', {
      body: { token: user.getJwtToken(), id: user.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(403);
    expect(body).toMatchSnapshot();
  });
});
