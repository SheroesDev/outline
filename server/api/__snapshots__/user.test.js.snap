// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#user.activate should activate a suspended user 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "createdAt": "2018-01-01T00:00:00.000Z",
    "email": "user1@example.com",
    "id": "46fde1d4-0050-428f-9f0b-0bf77f4bdf61",
    "isAdmin": false,
    "isSuspended": false,
    "name": "User 1",
    "username": "user1",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`#user.activate should require admin 1`] = `
Object {
  "error": "admin_required",
  "message": "An admin role is required to access this resource",
  "ok": false,
  "status": 403,
}
`;

exports[`#user.delete should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#user.demote should demote an admin 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "createdAt": "2018-01-01T00:00:00.000Z",
    "email": "user1@example.com",
    "id": "46fde1d4-0050-428f-9f0b-0bf77f4bdf61",
    "isAdmin": false,
    "isSuspended": false,
    "name": "User 1",
    "username": "user1",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`#user.demote should require admin 1`] = `
Object {
  "error": "admin_required",
  "message": "An admin role is required to access this resource",
  "ok": false,
  "status": 403,
}
`;

exports[`#user.demote shouldn't demote admins if only one available  1`] = `
Object {
  "error": "validation_error",
  "message": "At least one admin is required",
  "ok": false,
  "status": 400,
}
`;

exports[`#user.promote should promote a new admin 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "createdAt": "2018-01-01T00:00:00.000Z",
    "email": "user1@example.com",
    "id": "46fde1d4-0050-428f-9f0b-0bf77f4bdf61",
    "isAdmin": true,
    "isSuspended": false,
    "name": "User 1",
    "username": "user1",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`#user.promote should require admin 1`] = `
Object {
  "error": "admin_required",
  "message": "An admin role is required to access this resource",
  "ok": false,
  "status": 403,
}
`;

exports[`#user.suspend should require admin 1`] = `
Object {
  "error": "admin_required",
  "message": "An admin role is required to access this resource",
  "ok": false,
  "status": 403,
}
`;

exports[`#user.suspend should suspend an user 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "createdAt": "2018-01-01T00:00:00.000Z",
    "email": "user1@example.com",
    "id": "46fde1d4-0050-428f-9f0b-0bf77f4bdf61",
    "isAdmin": false,
    "isSuspended": true,
    "name": "User 1",
    "username": "user1",
  },
  "ok": true,
  "status": 200,
}
`;

exports[`#user.suspend shouldn't allow suspending the user themselves 1`] = `
Object {
  "error": "validation_error",
  "message": "Unable to suspend the current user",
  "ok": false,
  "status": 400,
}
`;

exports[`#user.update should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#user.update should update user profile information 1`] = `
Object {
  "data": Object {
    "avatarUrl": "http://example.com/avatar.png",
    "createdAt": "2018-01-01T00:00:00.000Z",
    "email": "user1@example.com",
    "id": "46fde1d4-0050-428f-9f0b-0bf77f4bdf61",
    "isAdmin": false,
    "isSuspended": false,
    "name": "New name",
    "username": "user1",
  },
  "ok": true,
  "status": 200,
}
`;
